_posts/2019-09-13-applying-code-splitting-aws-lambda.md

```diff
@@ -0,0 +1,180 @@
+
+A deletion handler for a particular resource-type then looks like this. Note that it only imports the client it requires, so that it doesn't need to load the whole aws-sdk package.
+
+```typescript
+import catalog from "aws-sdk/clients/servicecatalog";
+import { SplitArn, Call } from "../lib";
+
+const re = new RegExp(/^product\/(?<ProductId>.+)/);
+
+export default async function DeleteServiceCatalogCloudFormationProduct(arn: string) {
+    const { resource } = SplitArn(arn);
+    const [ _, Id ] = re.exec(resource);
+    // Here I have wrapped the API call to ignore deletion errors from 
+    // resources that do not exist
+    await Call(() => new catalog().deleteProduct({ Id }));
+}
+```
+
+This does some take some additional configuring of webpack to get this all working. I had been using the default webpack configuration producted by Rich Buggy [here](https://github.com/SnappyTutorials/aws-sam-webpack-plugin), but found it caused the "split-code" to be output in the wrong location. This meant that it was failing to be uploaded to S3 correctly. I still use Rich's plugin, but I modified the configuration to use webpack in multi-compiler mode, and this fixed the issue. This was my final webpack configuration:
+
+```js
+const AwsSamPlugin = require("aws-sam-webpack-plugin");
+const awsSamPlugin = new AwsSamPlugin();
+const path = require("path");
+
+module.exports = Object.entries(awsSamPlugin.entry()).map(([name, entry]) => {
+  return {
+    entry,
+    output: {
+      filename: "app.js",
+      libraryTarget: "commonjs2",
+      chunkFilename: "[id].js",
+      path: path.join(__dirname, `/.aws-sam/build/${name}`)
+    },
+    devtool: false,
+    resolve: {
+      extensions: [".ts", ".js"]
+    },
+    target: "node",
+    externals: [
+      "aws-sdk", 
+      /aws-sdk\/clients\/.*/
+    ],
+    mode: process.env.NODE_ENV || "production",
+    optimization: {
+      minimize: false
+    },
+    module: {
+      rules: [
+        {
+          test: /\.tsx?$/,
+          loader: "babel-loader"
+        }
+      ]
+    },
+    plugins: [
+      awsSamPlugin
+    ]
+  }
+})
+```
+
+Executing a build command results in the following;
+```
+└── build
+    ├── DeleteResources
+    │   ├── 0.js
+    │   ├── 1.js
+    │   ├── 2.js
+    │   └── app.js
+```
+
+The `app.js` file contains the entry point code, and each specific resource deletion function has been compiled into its own file identified by a digit. At runtime, the handler will dynamically import the code if it is required from the appropriate file. If the particular function is not required, it will never get resolved.
+
+I don't expect that this will likely bring about massive cost savings, but it was an interesting experiment to see if it was possible to implement code-splitting and dynamic resolution of modules at runtime in AWS Lambda.
```

Comment:
Are you able to provide any benchmarking - running the lambda before/after the code splitting?

Reply:
I could but it would be additional effort - I know Yan Cui has a set of step functions to benchmark stuff like this, so I'm somewhat hoping he takes the bait to do a more rigorous analysis and link back to this.